Add networkoption to disable non-TLS connections

sfc-gh-sgwydir opened the pull request:
By default, FDB clients configured with TLS credentials (CA, cert, key) are able to connect to clusters using TLS (and matching credentials) or plaintext clusters. As a potential attack vector, an attacker may be able to redirect a TLS-configured client's connections to a plaintext cluster in their control and read sensitive data. For security purposes, it is desirable to disallow a client to connect to non-TLS FDB clusters with a network option.

To solve this, I've added `NetworkOption::TLS_DISABLE_PLAINTEXT_CONNECTION` to fail all non-TLS connections made through `Net2::connect()` with a `connection_failed` error. 

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

sfc-gh-ajbeamon commented on flow/include/flow/TLSConfig.actor.h:
```diff
@@ -251,15 +260,16 @@ class TLSPolicy : ReferenceCounted<TLSPolicy> {
-	}
+	    { TLSConfig::OPT_TLS_PASSWORD, TLS_PASSWORD_FLAG, SO_REQ_SEP },                                                \
+		{ TLSConfig::OPT_TLS_CA_FILE, TLS_CA_FILE_FLAG, SO_REQ_SEP},                                                   \
+		{ TLSConfig::OPT_TLS_DISABLE_PLAINTEXT_CONNECTION, TLS_DISABLE_PLAINTEXT_CONNECTION_FLAG, SO_NONE }
```
Did something weird happen to the whitespace here?

sfc-gh-ajbeamon reviewed (commented):
Overall this looks good to me. Is there a way to prove by test that the option is working as intended?

sfc-gh-vgasiunas reviewed (commented):
It would be good to add some tests for the new options: at least one positive and one negative with the option enabled.

sfc-gh-sgwydir commented:
> It would be good to add some tests for the new options: at least one positive and one negative with the option enabled. One way to do this is fdb_c_client_config_tests.py.

Awesome -- what's the easiest way to run these tests?

sfc-gh-vgasiunas commented on bindings/c/test/fdb_c_client_config_tests.py:
```diff
@@ -267,6 +268,9 @@ def exec(self):
         if self.disable_client_bypass:
             cmd_args += ["--network-option-disable_client_bypass", ""]
 
+        if self.disable_client_bypass:
```
=> tls_disable_plaintext_connection

sfc-gh-vgasiunas commented on bindings/c/test/fdb_c_client_config_tests.py:
```diff
@@ -267,6 +268,9 @@ def exec(self):
             cmd_args += ["--network-option-disable_client_bypass", ""]
 
+        if self.disable_client_bypass:
+            cmd_args += ["--tls-disable-plaintext-connection", ""]
```
=> --network-option-tls_disable_plaintext_connection

sfc-gh-vgasiunas commented on bindings/c/test/client_config_tester.cpp:
```diff
@@ -92,6 +92,7 @@ class TesterOptions {
 	std::string externalClientDir;
 	bool disableLocalClient = false;
 	bool disableClientBypass = false;
+	bool tlsDisablePlainTextConnection = false;
```
Don't need to add anything, because now you can pass any network option using the syntax --network-option-[option_name]

sfc-gh-vgasiunas commented:
You can run the test with:
devDocker
ctest -R fdb_c_client_config_tests -VV

For debugging, it is also possible to execute fdb_c_client_config_tests.py directly and then filter by individual unit tests, but it is a bit more tricky, because you need to set up all Python dependencies it needs, paths and packages. I can help you with that if you need it.

sfc-gh-vgasiunas commented:
I see the CI is failing because of wrong formatting

sfc-gh-jshim commented on flow/TLSConfig.actor.cpp:
```diff
@@ -64,6 +65,15 @@ std::vector<std::string> LoadedTLSConfig::getVerifyPeers() const {
+bool LoadedTLSConfig::getDisablePlainTextConnection() const {
+	std::string envDisablePlainTextConnection;
+	if (platform::getEnvironmentVar("FDB_TLS_DISABLE_PLAINTEXT_CONNECTION", envDisablePlainTextConnection)) {
+		return boost::lexical_cast<bool>(envDisablePlainTextConnection);
```
This could throw a [bad_lexical_cast](https://www.boost.org/doc/libs/1_82_0/doc/html/boost_lexical_cast/synopsis.html#boost_lexical_cast.synopsis.bad_lexical_cast) and we don't catch those in FDB/fdbcli, so it would crash the process.
It could be more graceful to catch and log/print the error.
Alternatively you could use `try_lexical_convert`.

sfc-gh-jshim commented on flow/TLSConfig.actor.cpp:
```diff
@@ -211,6 +221,15 @@ std::string TLSConfig::getCAPathSync() const {
+bool TLSConfig::getDisablePlainTextConnection() const {
+	std::string envDisablePlainTextConnection;
+	if (platform::getEnvironmentVar("FDB_TLS_DISABLE_PLAINTEXT_CONNECTION", envDisablePlainTextConnection)) {
+		return boost::lexical_cast<bool>(envDisablePlainTextConnection);
```
I don't see why we should read EnvVar in both places (`LoadedTLSConfig` and `TLSConfig`) seeing as this variable doesn't change for the duration of the process. Loading it once in `TLSConfig` and forwarding it to `LoadedTLSConfig` should be enough. `LoadedTLSConfig` can refer to the cached boolean.

sfc-gh-jshim commented on fdbclient/vexillographer/fdb.options:
```diff
@@ -104,6 +104,8 @@ description is not currently required but encouraged.
             paramType="String" paramDescription="key passphrase"
             description="Set the passphrase for encrypted private key. Password should be set before setting the key for the password to be used." />
+    <Option name="TLS_disable_plaintext_connection" code="55"
+            description="Disable non-TLS connections" />
```
It could be more descriptive: e.g. `Prevent client from connecting to a non-TLS endpoint by throwing an error`.

sfc-gh-vgasiunas commented:
The build failing on:
/codebuild/output/src559314577/src/github.com/apple/foundationdb/documentation/sphinx/source/api-python.rst:211:Undefined substitution referenced: "option-set-disable-plaintext-connection".

sfc-gh-vgasiunas commented on bindings/c/test/fdb_c_client_config_tests.py:
```diff
@@ -338,6 +342,16 @@ def setUpClass(cls):
+        test = ClientConfigTest(self)
+        test.print_status = True
+        test.tls_disable_plaintext_connection = True
+        test.transaction_timeout = 100
```
Are we expected to fail with timeouts on TLS configuration errors? @sfc-gh-jshim do you know?

sfc-gh-jshim replied:
Typically a failed [connection attempt](https://github.com/apple/foundationdb/blob/main/fdbrpc/FlowTransport.actor.cpp#L702) would result in [retry with backoff](https://github.com/apple/foundationdb/blob/a7cf82adb2294f2468cb79c4b5962da040f89f59/fdbrpc/FlowTransport.actor.cpp#L687).
But now that you mention it, we could do better here.
Whether an endpoint is TLS is known in advance to connection attempt: the `:tls` address suffix.
Instead of re-using `connection_failed`, suffering the wait until timeout, and giving the false impression of some server or infrastructure failure, we could let it fail early on with a different error code: e.g. `plaintext_connection_prevented`. Gatekeeping function could be `TransportData::getOrOpenPeer()`, comparing the destination address' `:tls` suffix with the global network option, only if `FlowTransport::isClient()`.

sfc-gh-vgasiunas replied:
If the TLS errors are retriable, then it is in line with what we have now. We can plan a separate task for improving error reporting on such errors

sfc-gh-sgwydir commented:
Added positive and negative tests with individually created clusters for TLS enabled clusters. Added `disable_server_side_tls` to test the case where a client is configured with TLS but is connecting to a plaintext cluster. 

Having trouble with `test_plaintext_cluster_tls_client_plaintext_connection_disabled` -- the client cannot connect and emits `WARNING: Long delay (Ctrl-C to interrupt)` but does not timeout as I'd expect.

sfc-gh-vgasiunas commented:
Please fix the build error, because otherwise the tests even do not get executed by the CI:
```
Warning, treated as error:
/codebuild/output/src405183633/src/github.com/apple/foundationdb/documentation/sphinx/source/api-python.rst:211:Undefined substitution referenced: "option-set-disable-plaintext-connection".
```

sfc-gh-sgwydir commented:
> The error `WARNING: Long delay (Ctrl-C to interrupt)` comes from `fdbcli`. It means the test fails to setup the cluster with the given configuration. It uses `fdbcli` to create the database. Possibly `fdbcli` is not able to connect to the cluster with the provided TLS configuration.

I see. This is expected since `fdbcli` is being prevented from connecting at all. Is there a way of asserting this failure?

sfc-gh-sgwydir commented:
> The error `WARNING: Long delay (Ctrl-C to interrupt)` comes from `fdbcli`. It means the test fails to setup the cluster with the given configuration. It uses `fdbcli` to create the database. Possibly `fdbcli` is not able to connect to the cluster with the provided TLS configuration.

This was a bug -- I was blocking plaintext connections in the server as well. Thanks for your help. Tests are passing now.

ammolitor commented:
> ### Result of foundationdb-pr on Linux CentOS 7
> * Commit ID: [37bd59e](https://github.com/apple/foundationdb/commit/37bd59e41a498f98c88ccb333831873d511486d2)
> * Duration 0:59:16
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: ninja -v -C build_output -j ${NPROC} all packages strip_targets. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/foundationdb/foundationdb-pr/21844f23-8127-4a0c-846e-d390311eb2e6.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/foundationdb-pr/21844f23-8127-4a0c-846e-d390311eb2e6/workspace.zip) zip file of the working directory (available for 30 days)

this failed with `distcc` errors, I've started another build.
